File: desktop/src/CookieImport.cpp
#include "CookieImport.h"

#include <QDebug>
#include <QDir>
#include <QFile>
#include <QFileInfo>
#include <QProcess>
#include <QRegularExpression>
#include <QStringList>

namespace {
const char *kNetscapeHeader = "# Netscape HTTP Cookie File";
const char *kComment = "# 由 Hyper Online Video 从系统浏览器导入";

}   // namespace

QString CookieImport::siteOfDomain(const QString &domainIn) {
    const QString d = domainIn.startsWith('.') ? domainIn.mid(1) : domainIn;
    const QString low = d.toLower();
    if (low.endsWith("youtube.com") || low.endsWith("google.com") || low.endsWith("googlevideo.com")
        || low.endsWith("youtu.be") || low.endsWith("googleapis.com") || low.endsWith("gstatic.com")) {
        return "youtube";
    }
    if (low.endsWith("bilibili.com") || low.endsWith("hdslb.com") || low.endsWith("b23.tv")
        || low.endsWith("bilivideo.com")) {
        return "bilibili";
    }
    if (low.endsWith("163.com") || low.endsWith("126.com") || low.endsWith("126.net")
        || low.endsWith("music.126.net")) {
        return "netease";
    }
    if (low.endsWith("qq.com") || low.endsWith("gtimg.cn") || low.endsWith("qqmusic.qq.com")
        || low.endsWith("y.qq.com")) {
        return "qqmusic";
    }
    return QString();
}

QHash<QString, QString> CookieImport::splitBySite(const QString &combined) {
    QHash<QString, QList<QString>> buckets;
    const QStringList lines = combined.split(QRegularExpression("\r?\n"));
    for (const QString &line : lines) {
        if (line.trimmed().isEmpty() || line.startsWith('#')) continue;
        const QStringList f = line.split('\t');
        if (f.size() < 7) continue;                       // 不是合法的 Netscape 行
        const QString site = siteOfDomain(f.at(0));
        if (site.isEmpty()) continue;                     // 与本应用无关的域名直接丢弃（隐私）
        const QString domain = f.at(0);
        // Netscape 第 2 列必须与域名是否带前导点一致（yt-dlp 否则报 invalid format）
        const QString includeSub = domain.startsWith('.') ? "TRUE" : "FALSE";
        const QString rebuilt = QStringList{ domain, includeSub, f.at(2), f.at(3), f.at(4), f.at(5), f.at(6) }
                                    .join('\t');
        buckets[site].append(rebuilt);
    }
    QHash<QString, QString> out;
    for (auto it = buckets.begin(); it != buckets.end(); ++it) {
        QString text = QString("%1\n%2\n\n").arg(kNetscapeHeader, kComment);
        text += it.value().join('\n');
        text += '\n';
        out.insert(it.key(), text);
    }
    return out;
}

QHash<QString, int> CookieImport::countsOf(const QHash<QString, QString> &bySite) {
    QHash<QString, int> out;
    for (auto it = bySite.begin(); it != bySite.end(); ++it) {
        int n = 0;
        for (const QString &l : it.value().split('\n'))
            if (!l.trimmed().isEmpty() && !l.startsWith('#')) n++;
        out.insert(it.key(), n);
    }
    return out;
}

bool CookieImport::importFromBrowser(const QString &browser, const QString &probeUrl, QString *summary) {
    if (browser.isEmpty()) {
        if (summary) *summary = "未指定浏览器";
        return false;
    }
    const QString dir = QDir::homePath() + "/.config/hov/cookies";
    QDir().mkpath(dir);
    const QString tmp = dir + "/.import-combined.txt";
    QFile::remove(tmp);

    // yt-dlp：从浏览器读 cookie，并把合并后的 jar 写回 tmp（这就是"导出"）
    QProcess p;
    p.start("yt-dlp", QStringList{ "--cookies-from-browser", browser, "--cookies", tmp,
                                   "--simulate", "--skip-download", "--no-warnings", probeUrl });
    if (!p.waitForStarted(5000) || !p.waitForFinished(90000)) {
        p.kill();
        if (summary) *summary = "yt-dlp 未能在 90 秒内完成（浏览器可能弹出了钥匙串授权窗口）";
        return false;
    }
    QFile f(tmp);
    if (!f.exists() || !f.open(QIODevice::ReadOnly)) {
        if (summary)
            *summary = QString("未能导出 cookie（yt-dlp 退出码 %1；浏览器名是否正确？）").arg(p.exitCode());
        return false;
    }
    const QString text = QString::fromUtf8(f.readAll());
    f.close();

    const QHash<QString, QString> bySite = splitBySite(text);
    if (bySite.isEmpty()) {
        QFile::remove(tmp);
        if (summary) *summary = "浏览器里没有找到与本站点相关的 cookie（请先在浏览器登录对应站点）";
        return false;
    }
    const QHash<QString, int> counts = countsOf(bySite);
    QStringList parts;
    for (auto it = bySite.begin(); it != bySite.end(); ++it) {
        const QString path = QString("%1/%2.txt").arg(dir, it.key());
        QFile out(path);
        if (out.open(QIODevice::WriteOnly | QIODevice::Truncate)) {
            out.write(it.value().toUtf8());
            out.close();
            parts << QString("%1=%2 条").arg(it.key()).arg(counts.value(it.key()));
        }
    }
    QFile::remove(tmp);                                   // 合并文件含全部浏览器 cookie → 立刻删掉
    if (summary) *summary = parts.isEmpty() ? "写入失败（目录不可写？）" : parts.join("，");
    return !parts.isEmpty();
}

File: macos/Sources/HyperOnlineVideo/CookieImport.swift
import Foundation

/// 从系统浏览器导入 cookie（A0 跨端统一登录；macOS 侧）
///
/// 原理与 Linux 端一致：yt-dlp 的 `--cookies FILE` 会**读入并把 cookie jar 写回**该文件，
/// 配合 `--cookies-from-browser` 就把浏览器 cookie 落成了 Netscape 文件（音乐 API 与 yt-dlp 都吃）。
/// 好处：不依赖内置 WebView（AppImage/Flatpak 同样可用），且绕开 Google 对嵌入式登录的限制。
enum CookieImport {
    /// 域名 → 站点
    static func siteOfDomain(_ domainIn: String) -> String {
        let d = domainIn.hasPrefix(".") ? String(domainIn.dropFirst()) : domainIn
        let low = d.lowercased()
        let rules: [(String, [String])] = [
            ("youtube", ["youtube.com", "google.com", "googlevideo.com", "youtu.be", "googleapis.com", "gstatic.com"]),
            ("bilibili", ["bilibili.com", "hdslb.com", "b23.tv", "bilivideo.com"]),
            ("netease", ["163.com", "126.com", "126.net"]),
            ("qqmusic", ["qq.com", "gtimg.cn"]),
        ]
        for (site, suffixes) in rules where suffixes.contains(where: { low.hasSuffix($0) }) { return site }
        return ""
    }

    /// 合并的 Netscape 内容 → 站点 → 该站点 Netscape 文本（含文件头；无关域名丢弃）
    static func splitBySite(_ combined: String) -> [String: String] {
        var buckets: [String: [String]] = [:]
        for line in combined.components(separatedBy: CharacterSet.newlines) {
            let t = line.trimmingCharacters(in: .whitespaces)
            if t.isEmpty || t.hasPrefix("#") { continue }
            let f = t.components(separatedBy: "\t")
            guard f.count >= 7 else { continue }
            let site = siteOfDomain(f[0])
            if site.isEmpty { continue }                       // 隐私：与本应用无关的域名直接丢弃
            // Netscape 第 2 列必须与域名是否带前导点一致（否则 yt-dlp 报 invalid format）
            let includeSub = f[0].hasPrefix(".") ? "TRUE" : "FALSE"
            buckets[site, default: []].append([f[0], includeSub, f[2], f[3], f[4], f[5], f[6]].joined(separator: "\t"))
        }
        var out: [String: String] = [:]
        for (site, lines) in buckets {
            out[site] = "# Netscape HTTP Cookie File\n# 由 Hyper Online Video 从系统浏览器导入\n\n" + lines.joined(separator: "\n") + "\n"
        }
        return out
    }

    static func countsOf(_ bySite: [String: String]) -> [String: Int] {
        var out: [String: Int] = [:]
        for (site, text) in bySite {
            out[site] = text.components(separatedBy: "\n").filter { !$0.trimmingCharacters(in: .whitespaces).isEmpty && !$0.hasPrefix("#") }.count
        }
        return out
    }

    /// 跑一次 yt-dlp（用浏览器 cookie）→ 合并 jar 落到临时文件 → 按站点切分写入 cookies/ 目录
    static func importFromBrowser(_ browser: String, probeUrl: String) -> (ok: Bool, summary: String) {
        guard !browser.isEmpty else { return (false, "未指定浏览器（设置项 network.cookiesFromBrowser）") }
        guard let ytdlp = UrlResolver.findExecutable("yt-dlp") else { return (false, "找不到 yt-dlp（brew install yt-dlp）") }
        let dir = Config.dir + "/cookies"
        try? FileManager.default.createDirectory(atPath: dir, withIntermediateDirectories: true)
        let tmp = dir + "/.import-combined.txt"
        try? FileManager.default.removeItem(atPath: tmp)

        let r = UrlResolver.runProcess(ytdlp, ["--cookies-from-browser", browser, "--cookies", tmp,
                                              "--simulate", "--skip-download", "--no-warnings", probeUrl], timeout: 120)
        guard let text = try? String(contentsOfFile: tmp, encoding: .utf8) else {
            let tail = r.err.components(separatedBy: "\n").filter { !$0.isEmpty }.suffix(2).joined(separator: " ")
            return (false, "未能导出 cookie（yt-dlp 退出码 \(r.code)）：\(tail)")
        }
        let bySite = splitBySite(text)
        guard !bySite.isEmpty else {
            try? FileManager.default.removeItem(atPath: tmp)
            return (false, "浏览器里没有与本站点相关的 cookie（请先在浏览器登录对应站点）")
        }
        let counts = countsOf(bySite)
        var parts: [String] = []
        for (site, body) in bySite {
            if (try? body.write(toFile: dir + "/\(site).txt", atomically: true, encoding: .utf8)) != nil {
                parts.append("\(site)=\(counts[site] ?? 0) 条")
            }
        }
        try? FileManager.default.removeItem(atPath: tmp)      // 合并文件含全部浏览器 cookie → 立刻删
        return (parts.isEmpty ? false : true, parts.isEmpty ? "写入失败（目录不可写？）" : parts.sorted().joined(separator: "，"))
    }
}
